File: 78-subsets/subsets.cpp
class Solution {
public:
    void allSubsequences(int index , vector<int> & nums , vector<int> &arr  ,vector<vector<int>> &result, int n ){
        // base case;
        if(index>=n){
            result.push_back(arr);
            return ;
        }
    // add the element 
    arr.push_back(nums[index]);
    allSubsequences(index+1 ,nums, arr, result , n);

    // Not adding the elemenet
    arr.pop_back();
    allSubsequences(index+1 , nums, arr, result , n);
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        // backtracking approach 
        vector<vector<int>> result;
        int index =0 ;
        int n = nums.size();
        vector<int> arr;
    allSubsequences(index, nums, arr, result ,n);
    return result;
    }
};

File: 78-subsets/subsets.java
class Solution {

    public void solve(int [] nums, int index , List<Integer> arr,List<List<Integer>> result){
        if(index>=nums.length){
            result.add(new ArrayList<>(arr));
            return ;
        }   
        arr.add(nums[index]);
        solve(nums, index+1 , arr, result);
        arr.remove(arr.size()-1);
        solve(nums , index +1 ,arr, result);
    }       

    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>>result = new ArrayList<>();
        List<Integer>arr = new ArrayList<>();
        solve(nums, 0 ,arr,result);
        return result;
    }
}